protocols/noise/src/protocol/x25519_spec.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! libp2p-spec compliant Noise protocols based on X25519.
+
+use crate::{NoiseConfig, NoiseError, Protocol, ProtocolParams};
+use libp2p_core::UpgradeInfo;
+use libp2p_core::identity;
+use rand::Rng;
+use x25519_dalek::{X25519_BASEPOINT_BYTES, x25519};
+use zeroize::Zeroize;
+
+use super::{*, x25519::X25519};
+
+const STATIC_KEY_DOMAIN: &str = "noise-libp2p-static-key:";
+
+/// A X25519 key.
+#[derive(Clone)]
+pub struct X25519Spec([u8; 32]);
+
+impl AsRef<[u8]> for X25519Spec {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl Zeroize for X25519Spec {
+    fn zeroize(&mut self) {
+        self.0.zeroize()
+    }
+}
+
+impl Keypair<X25519Spec> {
+    /// Create a new X25519 keypair.
+    pub fn new() -> Keypair<X25519Spec> {
+        let mut sk_bytes = [0u8; 32];
+        rand::thread_rng().fill(&mut sk_bytes);
+        let sk = SecretKey(X25519Spec(sk_bytes)); // Copy
+        sk_bytes.zeroize();
+        Self::from(sk)
+    }
+}
+
+/// Promote a X25519 secret key into a keypair.
+impl From<SecretKey<X25519Spec>> for Keypair<X25519Spec> {
+    fn from(secret: SecretKey<X25519Spec>) -> Keypair<X25519Spec> {
+        let public = PublicKey(X25519Spec(x25519((secret.0).0, X25519_BASEPOINT_BYTES)));
+        Keypair { secret, public }
+    }
+}
+
+impl UpgradeInfo for NoiseConfig<XX, X25519Spec> {
+    type Info = &'static [u8];
+    type InfoIter = std::iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        std::iter::once(b"/noise")
```

Comment:
Is this the official protocol name?
If so, the fact that it is different from the existing one means that the upgrade path for Substrate/Polkadot will be easy.

Reply:
Yes, it's a bit hidden in the spec: https://github.com/libp2p/specs/tree/master/noise#negotiation.

I can only guess that https://github.com/libp2p/rust-libp2p/pull/1481 did not run into a protocol name issue because of doing only prefix matching on protocol names, but I really don't know.